source/ngf-mtl/impl.cpp

```diff
@@ -448,6 +465,64 @@ static uint32_t ngfmtl_get_num_rows(const uint32_t height, const ngf_image_forma
 
 #pragma mark ngf_struct_definitions
 
+enum ngf_id_init_types { id_default };
+
+// Shared pointer for managed objects
+template<typename T> class ngf_id {
+  public:
+  // Create an ngf_id with an additional retain count
+  // Useful for keeping AutoReleasePool managed objects alive beyond
+  // Their pool lifetime
+  static ngf_id add_retain(T* ptr) {
+    ngf_id res = ptr;
+    if (res) { res->retain(); }
+    return res;
+  }
+
+  ngf_id() : ptr_(nullptr) {
+  }
+  ngf_id(const ngf_id_init_types& type) : ptr_(T::alloc()->init()) {
+  }
+  // Note: Does NOT increment ref count. You can use this directly after calling alloc()->init()
+  ngf_id(T* starting_ptr) : ptr_(starting_ptr) {
+  }
+  ~ngf_id() {
+    if (ptr_) { ptr_->release(); }
+  }
+
+  ngf_id(const ngf_id& other) : ptr_(other.ptr_) {
+    if (ptr_) { ptr_->retain(); }
```

Comment:
i prefer to do `*this = other` (and `*this = std::move(other)`  for move ctor)

Reply:
Could you explain at a conceptual level why the `std::move` is necessary in the move constructor? Isn't the `other` variable already an 'expiring value' (ie, `&&')?